src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -1972,15 +1972,170 @@ public void observe(@NonNull SCMHead head, @NonNull SCMRevision revision) throws
+
+                    needSave = true;
+                    // get the previous seen revision
+                    SCMRevision scmLastSeenRevision = getLastSeenRevision(project, scmLastBuiltRevision);
+
+                    doAutomaticBuilds(head, revision, rawName, project, scmLastBuiltRevision, scmLastSeenRevision);
+
+                } else {
+                    listener.getLogger().format("No changes detected: %s (still at %s)%n", rawName, revision);
+                }
+
+            }
+
+            try {
+                if (needSave) {
+                    project.save();
+                }
+            } catch (IOException e) {
+                printStackTrace(e, listener.error("Could not save changes to " + rawName));
+            }
+        }
+
+        private void observeNew(@NonNull SCMHead head, @NonNull SCMRevision revision, @NonNull Branch branch, String rawName, String encodedName) {
+            P project;
+            if (!observer.mayCreate(encodedName)) {
+                listener.getLogger().println("Ignoring duplicate branch project " + rawName);
+                return;
+            }
+            try (ChildNameGenerator.Trace trace = ChildNameGenerator.beforeCreateItem(
+                    MultiBranchProject.this, encodedName, branch.getName()
+            )) {
+                if (getItem(encodedName) != null) {
+                    throw new IllegalStateException(
+                            "JENKINS-42511: attempted to redundantly create " + encodedName + " in "
+                                    + MultiBranchProject.this);
+                }
+                project = _factory.newInstance(branch);
+            }
+            if (!project.getName().equals(encodedName)) {
+                throw new IllegalStateException(
+                        "Name of created project " + project + " did not match expected " + encodedName);
+            }
+            // HACK ALERT
+            // ==========
+            // We don't want to trigger a save, so we will do some trickery to ensure that observer.created(project)
+            // performs the save
+            BulkChange bc = new BulkChange(project);
+            try {
+                if (project.getDisplayNameOrNull() == null && !rawName.equals(encodedName)) {
+                    project.setDisplayName(rawName);
+                }
+            } catch (IOException e) {
+                // ignore even if it does happen we didn't want a save
+            } finally {
+                bc.abort();
+            }
+            // decorate contract is to ensure it dowes not trigger a save
+            _factory.decorate(project);
+            // ok it is now up to the observer to ensure it does the actual save.
+            observer.created(project);
+            doAutomaticBuilds(head, revision, rawName, project, null, null);
+        }
+
+        private boolean isChangesDetected(@NonNull SCMRevision revision, @NonNull P project, SCMRevision scmLastBuiltRevision) {
```

Comment:
That singular `is` mixed with the plural of `changes` burns my eyes a bit :D
```suggestion
        private boolean changesDetected(@NonNull SCMRevision revision, @NonNull P project, SCMRevision scmLastBuiltRevision) {
```